### PR TITLE
ImportC: run cpp on Posix .c files

### PR DESCRIPTION
**Pull request by WalterBright**
It's way past time for dmd to start running the C preprocessor. This kicks that baby bird out of the nest so it can learn to fly.

I like incrementalism, so this PR does as little as possible:

1. it only works on .c files
2. if you don't want to run the preprocessor on Posix, name the file with a .i extension
3. it runs `cpp` on the default path
4. no parameters can be passed to `cpp`
5. doesn't automatically `#include <importc.h>`
6. error messages are minimal
7. most of the code is copypasta from `runProgram()` in `link.d`

Hence review should be easy.

Based on Max's prior work https://github.com/maxhaton/dmd/commit/e132151a02c6b7434544207321cd084145c780b5 greatly simplified.

### CODE REVIEW

**Comment by WalterBright** [comment]
Eh, looks like lexer.d is not liking the output of `cpp`. Will look at that tomorrow.

File: src/dmd/dmodule.d
Comment (ibuclaw) [inline src/dmd/dmodule.d:674]: No.

File: src/dmd/cpreprocess.d
Comment (ibuclaw) [inline src/dmd/cpreprocess.d:21]: Because, no dmd driver/backend code should be pulled into the front-end. (cc @kinke).

**Review by ibuclaw — DISMISSED** [review]

File: src/dmd/cpreprocess.d
Comment (WalterBright) [reply src/dmd/cpreprocess.d:21]: offer a suggestion

**Comment by WalterBright** [comment]
@ibuclaw this fails on linux with:
```
... dshell/issue22816.d
==============================
Test 'dshell/issue22816.d' failed. The logged output:
[COMPILE_TEST] '/tmp/cirrus-ci-build/generated/linux/release/64/dmd' '-conf=' '-m64' '-fPIC' '-odtest_results/dshell/issue22816' '-oftest_results/dshell/issue22816/run' '-I=/tmp/cirrus-ci-build/test/dshell' '-I=test_results/dshell/issue22816' '-I=/tmp/cirrus-ci-build/test/tools/dshell_prebuilt' '-i' '-i=-dshell_prebuilt' 'test_results/dshell_prebuilt.o' '/tmp/cirrus-ci-build/test/dshell/issue22816.d'
[RUN_TEST] 'test_results/dshell/issue22816/run'
[RUN] '/tmp/cirrus-ci-build/generated/linux/release/64/dmd' '-m64' '-c' 'dshell/extra-files/issue22816.c'
core.exception.AssertError@/tmp/cirrus-ci-build/test/tools/dshell_prebuilt/dshell_prebuilt.d(287): Expected 'Error: cannot find input file'
```
The failing test case https://github.com/dlang/dmd/pull/13713 is written by you and I don't know what it is supposed to be doing. Please help.

**Comment by ibuclaw** [comment]
> @ibuclaw this fails on linux with:
> ```
> ... dshell/issue22816.d
> ==============================
> Test 'dshell/issue22816.d' failed. The logged output:
> [COMPILE_TEST] '/tmp/cirrus-ci-build/generated/linux/release/64/dmd' '-conf=' '-m64' '-fPIC' '-odtest_results/dshell/issue22816' '-oftest_results/dshell/issue22816/run' '-I=/tmp/cirrus-ci-build/test/dshell' '-I=test_results/dshell/issue22816' '-I=/tmp/cirrus-ci-build/test/tools/dshell_prebuilt' '-i' '-i=-dshell_prebuilt' 'test_results/dshell_prebuilt.o' '/tmp/cirrus-ci-build/test/dshell/issue22816.d'
> [RUN_TEST] 'test_results/dshell/issue22816/run'
> [RUN] '/tmp/cirrus-ci-build/generated/linux/release/64/dmd' '-m64' '-c' 'dshell/extra-files/issue22816.c'
> core.exception.AssertError@/tmp/cirrus-ci-build/test/tools/dshell_prebuilt/dshell_prebuilt.d(287): Expected 'Error: cannot find input file'
> ```
> The failing test case https://github.com/dlang/dmd/pull/13713 is written by you and I don't know what it is supposed to be doing. Please help.

See #13707 for the fix.

It is expecting an error due to filemanager.lookup failing, but now you are calling preprocess and possibly getting another error (I hope not success).

To reproduce/get the new error message:
```
touch issue22816.d
dmd issue22816.c
```
Yes, those file extensions are correct.

**Comment by WalterBright** [comment]
@ibuclaw thanks for the quick answer

File: src/dmd/README.md
Comment (WalterBright) [reply src/dmd/README.md:177]: I didn't in order to minimize diffs for review purposes.

Comment (thewilsonator) [reply src/dmd/README.md:177]: that's what the ignore whitespace feature of the review screen is for (click the "cog" icon and select "Hide Whitespace" checkbox).

File: src/dmd/globals.d
Comment (ibuclaw) [inline src/dmd/globals.d:338]: Having the function `extern(D)` is not much use for anyone overriding this.

File: src/dmd/dmodule.d
Comment (ibuclaw) [inline src/dmd/dmodule.d:680]: Check for `!is null`.

File: src/dmd/cpreprocess.d
Comment (ibuclaw) [reply src/dmd/cpreprocess.d:21]: The Compiler interface might be one place for it. That is at least one module that is not meant to be common amongst implementing compilers.

**Review by ibuclaw — CHANGES_REQUESTED** [review]

File: src/dmd/dmodule.d
Comment (WalterBright) [reply src/dmd/dmodule.d:680]: It returns a struct, not a pointer.

File: src/dmd/globals.d
Comment (WalterBright) [reply src/dmd/globals.d:338]: I have no idea what you need for an interface or what your plans are for it, so I will leave that up to you.

**Comment by WalterBright** [comment]
We've had ImportC without preprocessor support for 6 months or thereabouts. It desperately needs to support the preprocessor. This is just a first stop. Please approve this so I can move forward to getting it completed and useful. We won't know what a better design would be without making more progress with this.

Nothing will move forward without this PR.

File: src/dmd/dmodule.d
Comment (thewilsonator) [reply src/dmd/dmodule.d:680]: I presume he means `if (global.preprocess !is null) filename = global.preprocess(srcfile);`?

File: src/dmd/root/env.d
Comment (thewilsonator) [inline src/dmd/root/env.d:23]: ?

File: src/dmd/utils.d
Comment (thewilsonator) [inline src/dmd/utils.d:22]: ?

File: src/dmd/README.md
Comment (thewilsonator) [inline src/dmd/README.md:178]: Realign `|`s on this line and the one above.

**Review by thewilsonator — DISMISSED** [review]

File: src/dmd/cpreprocess.d
Comment (thewilsonator) [inline src/dmd/cpreprocess.d:50]: We should give a better error message here. If we can't extract the error message from `cpp` we should at least tell the user which file failed preprocessing (this was meant to be part of the previous review)

File: src/dmd/utils.d
Comment (WalterBright) [reply src/dmd/utils.d:22]: Because I was trying to make preprocess() nothrow. I failed due to other problems, but the nothrows should still stick where they can.

Comment (thewilsonator) [reply src/dmd/utils.d:22]: They should go in a different PR then.

File: src/dmd/root/env.d
Comment (WalterBright) [reply src/dmd/root/env.d:23]: > I presume he means

I prefer to ask @ibuclaw to explain, as I tire of guessing :-)

File: src/dmd/cpreprocess.d
Comment (thewilsonator) [reply src/dmd/cpreprocess.d:50]: Fair. Change that to `error(Loc.initial, "C preprocess failed for file %s\n", csrcfile.toChars());` then.

File: src/dmd/dmodule.d
Comment (ibuclaw) [reply src/dmd/dmodule.d:680]: > I prefer to ask @ibuclaw to explain, as I tire of guessing :-)

Yes, the function pointer might be null.

File: src/dmd/cpreprocess.d
Comment (WalterBright) [reply src/dmd/cpreprocess.d:46]: Yes there is. But:
1. First of all, I want to get something to work, with minimal diffs.
2. One thing at a time.
3. I expect the design to evolve as we figure out what works best.
4. As mentioned, I have no idea what you have in mind for making it work with gdc.
5. I have mentioned using dmpp for cross compilation, which elicited no interest.

File: src/dmd/mars.d
Comment (WalterBright) [reply src/dmd/mars.d:733]: It never changes, so it is not necessary. Also do not want to need to import cpreprocess for that function.

Comment (ljmf00) [reply src/dmd/mars.d:733]: I mean, for `libdmd`, the `initDMD()` function at `frontend.d`. This file is driver-specific.

File: src/dmd/dmodule.d
Comment (ljmf00) [reply src/dmd/dmodule.d:680]: > I expect a "no-op preprocess" function would be provided that simply returned its argument. It's more attractive than adding the extra test.

If you want that, `global.preprocess` should have a "contract" check for non-null function pointers. I agree with @ibuclaw anyway, because that way, there is a way to disable preprocess for those files.

Comment (WalterBright) [reply src/dmd/dmodule.d:680]: Don't need a contract check for null pointers, as the hardware does that for you.

Comment (WalterBright) [reply src/dmd/dmodule.d:680]: > it's counter-intuitive

It's a different style, called branchless programming. The nice thing about it is it removes clutter from the code.

> and a function call has higher latency for that purpose, instead of a simple test check.

I'd agree if it was an inner loop, but it's once per file. Not significant.

File: src/dmd/mars.d
Comment (WalterBright) [reply src/dmd/mars.d:733]: I expect the interface for preprocess() to evolve, this is not the final way it's going to be. Spending a lot of effort on drivers is not going to be very productive.

File: src/dmd/dmodule.d
Comment (ibuclaw) [reply src/dmd/dmodule.d:680]: Initially this would be unimplemented in gdc and ldc though, resulting in a segfault. If this field was default initialized with a noop lambda then that'd be different.

File: src/dmd/globals.d
Comment (ibuclaw) [inline src/dmd/globals.d:338]: This should be in global.h too.

File: src/dmd/mars.d
Comment (ljmf00) [reply src/dmd/mars.d:733]: It's literally adding the same line in that function. I can do that on a subsequent PR anyway.

File: src/dmd/dmodule.d
Comment (ibuclaw) [inline src/dmd/dmodule.d:675]: Posix, also it'd be good to move this version condition to dmd.mars where you set preprocess. As there is no reason why gdc (and ldc?) can't do this on Windows.

**Review by ibuclaw — DISMISSED** [review]

File: src/dmd/mars.d
Comment (ibuclaw) [inline src/dmd/mars.d:733]:
```suggestion
    version (Posix)
        global.preprocess = &preprocess;
```

File: src/dmd/dmodule.d
Comment (ibuclaw) [inline src/dmd/dmodule.d:675]:
```suggestion
        if (global.preprocess)
```

Comment (WalterBright) [reply src/dmd/dmodule.d:675]: I don't really know what it will take to make things work for gdc and ldc. My focus here is getting *something* to work.

**Comment by WalterBright** [comment]
This needs `sppn.exe` to run the Win32 tests. It's available at http://ftp.digitalmars.com/sppn.zip

**Comment by WalterBright** [comment]
Since there's been nothing happening with this PR, I decided to add in support for FreeBSD, OSX and Windows.

**Comment by WalterBright** [comment]
This is now working well for Posix systems. For non-Posix systems:

1. For Microsoft's preprocessor, it insists on printing the filename it's preprocessing, which causes all the fail_compilation tests to fail. Yes, even with /nologo. cl.exe does not appear to have a "quiet" mode.
2. For Win32 with Digital Mars, I'm pretty sure these are all failing because sppn.exe, the DMC preprocessor, is not on the path. Even if it was, it would still fail because the DMC headers are not there. To get the headers and sppn.exe, dmc will need to be installed on the target machines: http://ftp.digitalmars.com/Digital_Mars_C++/Patch/dm857c.zip

File: test/fail_compilation/test22899.c
Comment (huglovefan) [inline test/fail_compilation/test22899.c:18]: this added brace adds a (second) syntax error on top of the one being tested

it should probably be moved to an `.i` file instead if one of the preprocessors complained about it

File: src/dmd/dmodule.d
Comment (ljmf00) [reply src/dmd/dmodule.d]: It's a step towards that. This is mixing driver/implementation-specific logic in `dmodule.d`, which should be generic to any handler and there is nothing that prevents it to run the preprocessor on other platforms.

File: test/fail_compilation/test22899.c
Comment (WalterBright) [reply test/fail_compilation/test22899.c:18]: I prefer it to work with the preprocessor.

File: src/dmd/mars.d
Comment (WalterBright) [reply src/dmd/mars.d]: ok

File: src/dmd/dmodule.d
Comment (WalterBright) [reply src/dmd/dmodule.d]: ok

**Comment by WalterBright** [comment]
ping @ibuclaw

**Comment by RazvanN7** [comment]
@WalterBright there is [this comment](https://github.com/dlang/dmd/pull/13955#discussion_r843772537) from @ibuclaw that has not been addressed.

**Comment by RazvanN7** [comment]
cc @maxhaton. What are your thoughts on this?

File: src/dmd/cpreprocess.d
Comment (WalterBright) [reply src/dmd/cpreprocess.d:21]: It's already been replaced with a pointer.

**Comment by WalterBright** [comment]
@RazvanN7 it had already been addressed.

**Review by thewilsonator — DISMISSED** [review]
> it had already been addressed.

No it hasn't, `dmd/cpreprocess.d` still imports `dmd/link.d`

**Comment by WalterBright** [comment]
> No it hasn't, dmd/cpreprocess.d still imports dmd/link.d

Right. But cpreprocess.d is only called via a pointer. It does not need to be imported. It is only imported by mars.d, which is not used by gdc and ldc.

**Comment by RazvanN7** [comment]
Added the "72h no objection -> merge" label.

**Comment by ljmf00** [comment]
This should have a changelog entry.

**Comment by maxhaton** [comment]
> Added the "72h no objection -> merge" label.

The approach is fairly crude but I don't mind merging this for now, as I can still work on mine. This PR cannot last because it can't pick the right preprocessor, but it's good enough for the immediate future.

File: src/dmd/cpreprocess.d
Comment (kinke) [reply src/dmd/cpreprocess.d:21]: What's imported from `dmd.link`?

**Comment by kinke** [comment]
Somewhat off-topic, regarding hardcoded `cpp` for Posix targets - I assume there are according switches for gcc/clang too? LDC has a `-gcc` option to explicitly select the C[++] compiler for linking (other option: `CC` env var); it'd be nice to be able to reuse that for C preprocessing (especially for cross-compilation, where `-gcc` is often set to something like `aarch64-linux-gnu-gcc`).

**Comment by ibuclaw** [comment]
> Somewhat off-topic, regarding hardcoded `cpp` for Posix targets - I assume there are according switches for gcc/clang too? LDC has a `-gcc` option to explicitly select the C[++] compiler for linking (other option: `CC` env var); it'd be nice to be able to reuse that for C preprocessing (especially for cross-compilation, where `-gcc` is often set to something like `aarch64-linux-gnu-gcc`).

Agreed on the hard coding of `cpp` in this PR. For me it would pretty much mean that I can no longer run a part of the DMD testsuite locally anymore (I have a weird set-up with different sysroots for various sorts of compilers, however I'm usually running on the fringe anyway, so do [regularly encounter periods](https://issues.dlang.org/show_bug.cgi?id=14956) when tests are broken for long stretches of time :-).

That the preprocessor handler is an `extern(C++)` callback though, that would mean you have the opportunity to implement it as you see fit.

I have reservations on this callback expecting a new file to be created though.  I might chance it to instead do this all in memory without needing to fork().

**Comment by maxhaton** [comment]
> > Somewhat off-topic, regarding hardcoded `cpp` for Posix targets - I assume there are according switches for gcc/clang too? LDC has a `-gcc` option to explicitly select the C[++] compiler for linking (other option: `CC` env var); it'd be nice to be able to reuse that for C preprocessing (especially for cross-compilation, where `-gcc` is often set to something like `aarch64-linux-gnu-gcc`).
> 
> Agreed on the hard coding of `cpp` in this PR. For me it would pretty much mean that I can no longer run a part of the DMD testsuite locally anymore (I have a weird set-up with different sysroots for various sorts of compilers, however I'm usually running on the fringe anyway, so do [regularly encounter periods](https://issues.dlang.org/show_bug.cgi?id=14956) when tests are broken for long stretches of time :-).
> 
> That the preprocessor handler is an `extern(C++)` callback though, that would mean you have the opportunity to implement it as you see fit.
> 
> I have reservations on this callback expecting a new file to be created though. I might chance it to instead do this all in memory without needing to fork().

The design in my PR is intended at least somewhat to automate finding the right preprocessor, or make it illegal to use the wrong one. It's beerconf so why not discuss it there I think

**Comment by WalterBright** [comment]
This PR is coming up on being a month old. I don't have any angst about replacing it with something better, but right now we don't have anything better. We need something that works, and this works.

**Comment by WalterBright** [comment]
@ibuclaw what's holding this up?

**Comment by maxhaton** [comment]
> @ibuclaw what's holding this up?

I can't see it so needs to implicitly ImportC.h for one otherwise it's useless in practice because no system headers will work.

**Comment by WalterBright** [comment]
I'm planning on implicitly #include-ing importc.h after this one is merged.

**Comment by WalterBright** [comment]
72 hours are up!

**Comment by kinke** [comment]
> I assume there are according switches for gcc/clang too? 

Answering myself: yes, both gcc and clang support `-E` for preprocessing only.

> I'm planning on implicitly #include-ing importc.h after this one is merged.

For gcc and clang, `-include importc.h` does exactly that.

**Comment by WalterBright** [comment]
@kinke yes. There's also a switch for both MSVC and DMC's preprocessors to do a `-include`.

**Pull request merged**
